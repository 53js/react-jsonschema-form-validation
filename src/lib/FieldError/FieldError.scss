.Jfv_FieldError {
	display: none;
	width: 100%;
	margin-top: .25rem;
	font-size: 80%;
	color: $color-red;

	&.isSubmitted {
		display: block;
		text-align: left;
	}
}
