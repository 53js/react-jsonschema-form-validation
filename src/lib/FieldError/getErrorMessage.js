/**
 * [getErrorMessage find the error message to display when a field is invalid]
 * @param  {Object} fieldError           [The AJV error object]
 * @param  {Object} errorMessages   	 [The complete list of error messages]
 * @return {string}                      [return the error message]
 *
 * Message displayed priority :
 *  1. The correponding error message passed by the FieldError component
 *  2. The correponding error message passed by the Form component or
 *  3. The default message passed by the Form component
 *  4. The AJV message if noting is passed
 */
const getErrorMessage = (fieldError, errorMessages) => {
	// customMessage = the errorMessage provided by the <FieldError> as props
	// or the message passed as props to the <Form> as props
	const customErrorMessage = errorMessages
		&& (errorMessages[fieldError.keyword] || errorMessages.defaultMessage);

	let message = customErrorMessage ? customErrorMessage(fieldError) : fieldError.message;

	/* istanbul ignore next */
	if (process.env.REACT_APP_JFV_DEBUG === 'true') {
		message += ` [#${fieldError.keyword}]`;
	}

	return message;
};

export default getErrorMessage;
