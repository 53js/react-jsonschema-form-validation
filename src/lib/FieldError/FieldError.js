import classnames from 'classnames';
import memoize from 'memoize-one';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { withFormContext } from '../Form/Context';
import getErrorMessage from './getErrorMessage';

class FieldError extends PureComponent {
	memoGetClassnames = memoize((className, isSubmitted, isTouched) => classnames(
		'Jfv_FieldError',
		className,
		{
			isSubmitted,
			isTouched,
		},
	))

	memoGetFieldErrorMessage = memoize((error, formErrorMessages, fieldErrorMessages) => {
		const errorMessages = { ...formErrorMessages, ...fieldErrorMessages };
		return getErrorMessage(error, errorMessages);
	})

	getClassnames = (form) => {
		const { className, name } = this.props;
		const { isFieldTouched, isSubmitted } = form;

		return this.memoGetClassnames(
			className,
			isSubmitted,
			isFieldTouched(name),
		);
	}

	getFieldErrorMessage = (error, form) => {
		const { errorMessages: fieldErrorMessages } = this.props;
		const { errorMessages: formErrorMessages } = form;
		return this.memoGetFieldErrorMessage(error, formErrorMessages, fieldErrorMessages);
	}

	render() {
		const {
			children,
			className,
			component: Component,
			errorMessages,
			name,
			...props
		} = this.props;

		return withFormContext((form) => {
			const fieldErrors = form.getFieldErrors(name);
			if (!fieldErrors.length) return null;

			return (
				<Component className={this.getClassnames(form)} {...props}>
					{
						children
						|| this.getFieldErrorMessage(fieldErrors[0], form)
					}
				</Component>
			);
		});
	}
}

FieldError.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	component: PropTypes.elementType,
	errorMessages: PropTypes.objectOf(PropTypes.func),
	name: PropTypes.string.isRequired,
};

FieldError.defaultProps = {
	children: null,
	component: 'div',
	errorMessages: null,
	className: '',
};

export default FieldError;
