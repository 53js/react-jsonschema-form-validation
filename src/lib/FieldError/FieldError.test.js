import React from 'react';
import { mount } from 'enzyme';

import FormContext from '../Form/Context';
import FieldError from './FieldError';

jest.mock('../Form/Context');

it('should match snapshot', () => {
	const context = {
		getFieldErrors: jest.fn(() => [{ keyword: 'bad' }]),
		isFieldTouched: jest.fn(),
	};

	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const fieldError = mount(<FieldError name="username" />);
	expect(fieldError).toMatchSnapshot();
});

it('should not be displayed if field has no error', () => {
	const fieldError = mount(<FieldError name="username" />);
	expect(fieldError.find(FormContext.Consumer).exists()).toBe(true);
});

it('should call error message of first error only if field has errors', () => {
	const context = {
		errorMessages: {
			bad1: jest.fn(),
			bad2: jest.fn(),
		},
		getFieldErrors: jest.fn(() => [{ keyword: 'bad1' }, { keyword: 'bad2' }]),
		isFieldTouched: jest.fn(),
	};

	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	mount(<FieldError name="username" />);
	expect(context.errorMessages.bad1).toHaveBeenCalled();
	expect(context.errorMessages.bad2).not.toHaveBeenCalled();
});

it('should allow to extend and override error messages defined in form', () => {
	const context = {
		errorMessages: {
			bad1: jest.fn(),
			bad2: jest.fn(),
		},
		getFieldErrors: jest.fn(() => [{ keyword: 'bad1' }, { keyword: 'bad2' }]),
		isFieldTouched: jest.fn(),
	};

	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const bad1Override = jest.fn();
	mount(
		<FieldError
			errorMessages={{ bad1: bad1Override }}
			name="username"
		/>,
	);
	expect(bad1Override).toHaveBeenCalled();
	expect(context.errorMessages.bad1).not.toHaveBeenCalled();

	context.getFieldErrors = jest.fn(() => [{ keyword: 'bad3' }]);
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const bad3 = jest.fn();
	mount(
		<FieldError
			errorMessages={{ bad3 }}
			name="username"
		/>,
	);
	expect(bad3).toHaveBeenCalled();
	expect(context.errorMessages.bad1).not.toHaveBeenCalled();
});

it('should render children if providen', () => {
	const context = {
		getFieldErrors: jest.fn(() => [{ keyword: 'bad1' }]),
		isFieldTouched: jest.fn(),
	};
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	const fieldError = mount(
		<FieldError name="username">
			<span id="message" />
		</FieldError>,
	);
	expect(fieldError.exists('#message')).toBe(true);
});

it('should add class isSubmitted if form is submitted', () => {
	const context = {
		getFieldErrors: jest.fn(() => [{ keyword: 'bad1' }]),
		isFieldTouched: jest.fn(),
	};
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	let fieldError = mount(<FieldError name="username" />);
	expect(fieldError.find('.Sjf_FieldError').hasClass('isSubmitted')).toBe(false);

	context.isSubmitted = true;
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	fieldError = mount(<FieldError name="username" />);
	expect(fieldError.find('.Sjf_FieldError').hasClass('isSubmitted')).toBe(true);
});

it('should add class isTouched if field is touched', () => {
	const context = {
		getFieldErrors: jest.fn(() => [{ keyword: 'bad1' }]),
		isFieldTouched: jest.fn(),
	};
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	let fieldError = mount(<FieldError name="username" />);
	expect(fieldError.find('.Sjf_FieldError').hasClass('isTouched')).toBe(false);

	context.isFieldTouched.mockImplementation(() => true);
	FormContext.Consumer.mockImplementationOnce(props => props.children(context));
	fieldError = mount(<FieldError name="username" />);
	expect(fieldError.find('.Sjf_FieldError').hasClass('isTouched')).toBe(true);
});
