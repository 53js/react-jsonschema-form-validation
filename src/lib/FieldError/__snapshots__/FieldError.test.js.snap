// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<FieldError
  className=""
  component="div"
  errorMessages={null}
  name="username"
>
  <mockConstructor>
    <div
      className="Jfv_FieldError"
    />
  </mockConstructor>
</FieldError>
`;
